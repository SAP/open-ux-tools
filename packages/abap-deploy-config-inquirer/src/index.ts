import { ToolsLogger, type Logger } from '@sap-ux/logger';
import { initI18n } from './i18n';
import { PromptState } from './prompts/prompt-state';
import { getAbapDeployConfigQuestions } from './prompts';
import LoggerHelper from './logger-helper';
import type { InquirerAdapter } from '@sap-ux/inquirer-common';
import type { AbapDeployConfigAnswers, AbapDeployConfigPromptOptions } from './types';

/**
 * Get the inquirer prompts for abap deploy config.
 *
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompts used to provide input for abap deploy config generation and a reference to the answers object which will be populated with the user's responses once `inquirer.prompt` returns
 */
async function getPrompts(
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<{ prompts: any[]; answers: Partial<AbapDeployConfigAnswers> }> {
    await initI18n();
    LoggerHelper.logger = logger ?? new ToolsLogger({ logPrefix: '@sap-ux/abap-deploy-config-inquirer' });

    return {
        prompts: await getAbapDeployConfigQuestions(promptOptions),
        // Return reference to derived answers object that will be populated with user responses (after prompting is complete)
        answers: PromptState.abapDeployConfig
    };
}

/**
 * Prompt for abap deploy config writer inputs.
 *
 * @param adapter - optionally provide references to a calling inquirer instance, this supports integration to Yeoman generators, for example
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompt answers
 */
async function prompt(
    adapter: InquirerAdapter,
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<AbapDeployConfigAnswers> {
    const abapDeployConfigPrompts = (await getPrompts(promptOptions, logger)).prompts;
    const answers = await adapter.prompt<AbapDeployConfigAnswers>(abapDeployConfigPrompts);

    // Add dervied service answers to the answers object
    Object.assign(answers, PromptState.abapDeployConfig);

    return answers;
}

export { getPrompts, prompt, type InquirerAdapter, type AbapDeployConfigAnswers, type AbapDeployConfigPromptOptions };
