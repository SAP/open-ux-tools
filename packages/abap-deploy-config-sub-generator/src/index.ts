export { getAbapQuestions } from './app/questions';
export { indexHtmlExists, showOverwriteQuestion } from './utils';
export { AbapDeployConfigOptions, DeployProjectType } from './app/types';
