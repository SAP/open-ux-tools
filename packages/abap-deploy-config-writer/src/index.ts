import { join } from 'path';
import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';
import cloneDeep from 'lodash/cloneDeep';
import { addPackageDevDependency, FileName, readUi5Yaml, updatePackageScript } from '@sap-ux/project-access';
import { getDeployConfig, updateBaseConfig } from './config';
import { addUi5Dependency, writeUi5RepositoryFiles } from './file';
import {
    BUILD_SCRIPT,
    DEPLOY_SCRIPT,
    RIMRAF,
    RIMRAF_VERSION,
    UI5_TASK_FLATTEN_LIB,
    UI5_TASK_FLATTEN_LIB_VERSION,
    UNDEPLOY_SCRIPT
} from './constants';
import type { DeployConfigOptions } from './types';
import type { Editor } from 'mem-fs-editor';
import type { AbapDeployConfig } from '@sap-ux/ui5-config';

/**
 * Writes the template to the memfs editor instance.
 *
 * @param basePath - the base path
 * @param abapDeployConfig - the abap deploy config
 * @param options - the deploy config options
 * @param fs - the memfs editor instance
 * @returns the updated memfs editor instance
 */
async function generate(
    basePath: string,
    abapDeployConfig: AbapDeployConfig,
    options?: DeployConfigOptions,
    fs?: Editor
): Promise<Editor> {
    const abapConfig = cloneDeep(abapDeployConfig);

    if (!fs) {
        fs = create(createStorage());
    }

    // base config
    const baseConfigFile = options?.baseFile ?? FileName.Ui5Yaml;
    const baseConfig = await readUi5Yaml(basePath, baseConfigFile, fs);
    const baseConfigPath = join(basePath, baseConfigFile);
    updateBaseConfig(baseConfigPath, baseConfig, fs);

    // deploy config
    const deployConfigFile = options?.deployFile ?? FileName.UI5DeployYaml;
    const deployFilePath = join(basePath, deployConfigFile);
    const deployConfig = await getDeployConfig(abapConfig, baseConfig);
    fs.write(deployFilePath, deployConfig.toString());

    // package.json
    // deploy script
    const deployScript = `${BUILD_SCRIPT} && ${DEPLOY_SCRIPT} --config ${deployConfigFile} && ${RIMRAF} archive.zip`;
    await updatePackageScript(basePath, 'deploy', deployScript, fs);
    // undeploy script
    const undeployScript = `${BUILD_SCRIPT} && ${UNDEPLOY_SCRIPT} --config ${deployConfigFile}`;
    await updatePackageScript(basePath, 'undeploy', undeployScript, fs);
    // test mode script
    const deployTestModeScript = `${BUILD_SCRIPT} && ${DEPLOY_SCRIPT} --config ${deployConfigFile} --testMode true`;
    await updatePackageScript(basePath, 'deploy-test', deployTestModeScript, fs);
    // dependencies
    await addPackageDevDependency(basePath, RIMRAF, RIMRAF_VERSION, fs);
    await addPackageDevDependency(basePath, UI5_TASK_FLATTEN_LIB, UI5_TASK_FLATTEN_LIB_VERSION, fs);
    addUi5Dependency(fs, basePath, UI5_TASK_FLATTEN_LIB);

    // ui5 repo files
    await writeUi5RepositoryFiles(fs, basePath);

    return fs;
}

export { generate };
