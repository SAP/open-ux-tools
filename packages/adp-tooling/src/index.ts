export { AdpPreviewConfig, AdpWriterConfig } from './types';
export * from './preview/adp-preview';
export { generate } from './writer';
export * from './base/prompt';
