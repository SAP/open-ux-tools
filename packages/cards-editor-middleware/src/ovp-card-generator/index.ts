import type { RequestHandler, Request, Response } from 'express';
import { Router } from 'express';
import { LogLevel, ToolsLogger, UI5ToolingTransport } from '@sap-ux/logger';
import type { MiddlewareParameters } from '@ui5/server';
import * as utils from '../common/utils';
import * as constants from '../common/constants';
import { join, dirname } from 'path';
import { existsSync, writeFileSync, mkdirSync, readFileSync } from 'fs';
import bodyParser from 'body-parser';
import os from 'os';

export interface Configuration {
    target: string;
    path: string;
    webapp?: string;
}

const defaultConfig: Configuration = {
    webapp: constants.WEBAPP,
    target: constants.WEBAPP_CARDS_TARGET,
    path: constants.EDITOR_CARD
};
const i18nConfig: Configuration = {
    target: constants.WEBAPP_I18N_TARGET,
    path: constants.EDITOR_I18N
};

module.exports = ({ options }: MiddlewareParameters<any>): RequestHandler => {
    const logger = new ToolsLogger({
        transports: [new UI5ToolingTransport({ moduleName: 'ovp-cards-generator' })],
        logLevel: options.configuration?.debug ? LogLevel.Debug : LogLevel.Info
    });
    const router = Router();
    const config = {
        ...defaultConfig,
        ...options.configuration
    };
    router.use(bodyParser.json());
    router.use(config.path, (req: Request, res: Response) => {
        try {
            const folder = join(config.target, dirname(req.path));
            const file = utils.prepareFileName(req.path);
            if (!existsSync(folder)) {
                mkdirSync(folder, { recursive: true });
            }
            const card = utils.prepareCardForSaving(req.body); // TODO: remove once done in OVP
            writeFileSync(join(folder, file), card);
            // edit manifest to update embeds property
            const manifestPath = constants.APP_MANIFEST_PATH;
            const manifestFile = readFileSync(manifestPath, 'utf8');
            const oManifest = JSON.parse(manifestFile);
            if (!oManifest['sap.app'].embeds) {
                oManifest['sap.app'].embeds = [];
            }
            const embedEntry = join('cards', dirname(req.path));
            if (!oManifest['sap.app'].embeds.includes(embedEntry)) {
                oManifest['sap.app'].embeds.push(embedEntry);
                writeFileSync(manifestPath, JSON.stringify(oManifest, null, 2));
            }
            res.status(201).send(`${join(folder, file)} created or updated.`);
        } catch (err) {
            res.status(500).send(`File could not be created/updated.`);
        }
    });

    router.use(i18nConfig.path, (req: Request, res: Response) => {
        try {
            // read manifest to get the i18n file path
            const manifestPath = constants.APP_MANIFEST_PATH;
            const manifestFile = readFileSync(manifestPath, 'utf8');
            const oManifest = JSON.parse(manifestFile);
            const i18nFilePath = oManifest['sap.app'].i18n || '/i18n/i18n.properties';
            const filePath = join('./webapp', i18nFilePath);
            const entries = req.body || [];
            const updatedEntries: { [key: number]: boolean } = {};
            const output: string[] = [];
            const lines = utils.traverseI18nProperties(
                filePath,
                (line: string, index: number, keyTemp?: string, valueTemp?: string) => {
                    const existingIndex: number =
                        valueTemp !== undefined ? entries.findIndex((entry: any) => entry.key === keyTemp) : -1;
                    if (existingIndex !== -1) {
                        const { key, value } = entries[existingIndex];
                        line = `${key}=${value}`;
                        updatedEntries[existingIndex] = true;
                    }
                    output.push(line);
                }
            );
            // check if file does not end with new line
            if (lines.length > 0 && lines[lines.length - 1].trim() && entries.length) {
                // If there no end line - add new gap line before new content
                output.push('');
            }
            for (const index in entries) {
                if (!updatedEntries[index as any]) {
                    const { comment, key, value } = entries[index];
                    // New i18n entry - add it at the end of file
                    if (comment) {
                        // Add comment only for new entry
                        output.push(`#${comment}`);
                    }
                    output.push(`${key}=${value}${os.EOL}`);
                }
            }
            writeFileSync(filePath, output.join(os.EOL), { encoding: 'utf8' });

            res.status(201).send(`i18n file updated.`);
        } catch (err) {
            console.log(err);
            res.status(500).send(`File could not be updated.`);
        }
    });

    router.get('/editor/listCards', (req: Request, res: Response) => {
        const directory = join(config.target);
        let aManifests: any = [];
        utils.getDirectoriesRecursive(directory).forEach((subdir) => {
            if (!existsSync(subdir)) {
                return;
            } else {
                aManifests = [...aManifests, ...utils.getAllManifests(subdir)];
            }
        });
        res.json(aManifests);
    });

    logger.info(
        `Listening for POST requests matching ${config.path}/<card_name> that will create files at ${config.target}/<card_name>.json.`
    );
    return router;
};
