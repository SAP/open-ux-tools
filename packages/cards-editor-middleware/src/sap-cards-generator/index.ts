import type { RequestHandler, Request, Response } from 'express';
import { Router } from 'express';
import * as utils from '../common/utils';
import * as constants from '../common/constants';
import { join, dirname } from 'path';
import { existsSync, writeFileSync, mkdirSync, readFileSync } from 'fs';
import bodyParser from 'body-parser';
import { CARD_TYPES } from '../common/constants';

import { LogLevel, ToolsLogger, UI5ToolingTransport } from '@sap-ux/logger';

import type { MiddlewareParameters } from '@ui5/server';
import { render } from 'ejs';

interface Configuration {
    storeCard?: string;
    webapp?: string;
    cardsPreviewMiddleware?: string;
}

const defaultConfig: Configuration = {
    webapp: constants.WEBAPP
};

const generatorConfig: Configuration = {
    cardsPreviewMiddleware: constants.CARDS_PREVIEW_MIDDLEWARE,
    storeCard: constants.STORE_CARD
};

module.exports = async ({ resources, options }: MiddlewareParameters<any>): Promise<RequestHandler> => {
    const _logger = new ToolsLogger({
        transports: [new UI5ToolingTransport({ moduleName: 'sap-cards-generator' })],
        logLevel: options.configuration?.debug ? LogLevel.Debug : LogLevel.Info
    });
    const router = Router();

    const manifest = await resources.rootProject.byPath('/manifest.json');
    if (!manifest) {
        throw new Error('No manifest.json found.');
    }

    const config = {
        ...defaultConfig,
        ...options.configuration
    };

    router.get('/test/flpGeneratorSandbox.html', async (_req, res: Response) => {
        const app = JSON.parse(await manifest.getString())['sap.app'];
        res.status(200).send(
            render(readFileSync(join(__dirname, '../../templates/editor.html'), 'utf-8'), {
                templateModel: {
                    appTitle: app.title || 'Card Editor Preview',
                    component: app.id
                }
            })
        );
    });

    router.use(bodyParser.json());

    router.use(generatorConfig.storeCard as string, (req: Request, res: Response) => {
        try {
            const floorplan = req.query.floorplan;
            const localPath = req.query.localPath;
            const fileName = req.query.fileName || 'manifest.json';
            const folder = join(config.webapp || '', dirname(localPath + '/' + fileName));
            const file = utils.prepareFileName(localPath + '/' + fileName);

            if (!existsSync(folder)) {
                mkdirSync(folder, { recursive: true });
            }
            const multipleCards = utils.prepareCardTypesForSaving(req.body);
            writeFileSync(join(folder, file), multipleCards.integration);
            writeFileSync(join(folder, 'adaptive-' + file), multipleCards.adaptive);

            const manifestPath = constants.APP_MANIFEST_PATH;
            const manifestFile = readFileSync(manifestPath, 'utf8');
            const oManifest = JSON.parse(manifestFile);

            if (!oManifest['sap.cards.ap']) {
                oManifest['sap.cards.ap'] = {};
            }
            const integrationCard = req.body.find((card: any) => card.type === CARD_TYPES.INTERGATION);
            const entitySet = integrationCard.entitySet;
            if (!oManifest['sap.cards.ap'].embeds) {
                oManifest['sap.cards.ap'].embeds = {};
            }
            if (!oManifest['sap.cards.ap'].embeds[floorplan]) {
                oManifest['sap.cards.ap'].embeds[floorplan] = {
                    'default': entitySet,
                    'manifests': {}
                };
            }
            if (!oManifest['sap.cards.ap'].embeds[floorplan]['default']) {
                oManifest['sap.cards.ap'].embeds[floorplan]['default'] = entitySet;
            }
            if (!oManifest['sap.cards.ap'].embeds[floorplan]['manifests']) {
                oManifest['sap.cards.ap'].embeds[floorplan]['manifests'] = {};
            }
            if (!oManifest['sap.cards.ap'].embeds[floorplan]['manifests'][entitySet]) {
                oManifest['sap.cards.ap'].embeds[floorplan]['manifests'][entitySet] = [
                    {
                        'localUri': folder
                    }
                ];
            }
            writeFileSync(manifestPath, JSON.stringify(oManifest, null, 2));
            res.status(201).send(`${join(folder, file)} and ${join(folder, 'adaptive-' + file)} created or updated.`);
        } catch (err) {
            res.status(500).send(`Files could not be created/updated.`);
        }
    });

    router.get(generatorConfig.cardsPreviewMiddleware as string, (_req: Request, res: Response) => {
        try {
            res.status(200)
                .contentType('text/javascript')
                .send(readFileSync(join(__dirname, './generatorRenderer.js'), 'utf-8'));
        } catch (err) {
            res.status(404).send(`File not found.`);
        }
    });

    return router;
};
