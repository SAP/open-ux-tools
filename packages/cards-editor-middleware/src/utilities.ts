import { promises } from 'fs';
import packageJson from '../package.json';

interface MultiCardsPayload {
    type: string;
    manifest: object;
}

interface MultiCardsStringManifest {
    integration: string;
    adaptive: string;
}

export interface I18nEntry {
    key: string;
    value: string;
    comment?: string;
}

export const CARD_TYPES = {
    INTERGATION: 'integration',
    ADAPTIVE: 'adaptive'
};

/**
 * Prepare the file name by adding .json extension if it is missing.
 *
 * @param path Path to the file
 * @returns returns the file name
 */
export function prepareFileName(path: string): string {
    const fileName = path.split('/')[path.split('/').length - 1];
    return fileName.endsWith('.json') ? fileName : `${fileName}.json`;
}

/**
 * Prepare card for saving by adding dtMiddleware parameter to the card manifest.
 *
 * @param card {object}
 * @returns {void}
 */
function prepareIntegrationCardForSaving(card: any): void {
    const version = packageJson.version;
    const insights = card?.['sap.insights'];
    if (!insights.versions) {
        insights.versions = {
            dtMiddleware: version
        };
    } else {
        insights.versions.dtMiddleware = version;
    }
}

/**
 * This function takes the payload from the editor and prepares the integration and adaptive cards for saving.
 *
 * @param aMultipleCards {MultiCardsPayload[]}
 * @returns {MultiCardsStringManifest} returns the integration and adaptive cards as strings
 */
export function prepareCardTypesForSaving(aMultipleCards: MultiCardsPayload[]): MultiCardsStringManifest {
    const integrationCard = aMultipleCards.find((card) => card.type === CARD_TYPES.INTERGATION) ?? {
        type: CARD_TYPES.INTERGATION,
        manifest: {}
    };
    const adaptiveCard = aMultipleCards.find((card) => card.type === CARD_TYPES.ADAPTIVE) ?? {
        type: CARD_TYPES.ADAPTIVE,
        manifest: {}
    };
    prepareIntegrationCardForSaving(integrationCard.manifest);
    return {
        integration: JSON.stringify(integrationCard.manifest, null, 2),
        adaptive: JSON.stringify(adaptiveCard.manifest, null, 2)
    };
}

/**
 * Traverse the i18n properties file and call the callback function for each property.
 *
 * @param path {string} - Path to the i18n properties file
 * @param entries {Array<I18nEntry>} - Array of entries to be updated
 * @returns {object} - Object containing the lines, updated entries and the output
 */
export async function traverseI18nProperties(path: string, entries: Array<I18nEntry>) {
    const i18nFile: string = await promises.readFile(path, 'utf8');
    const lines = i18nFile.split(/\r\n|\n/);
    const updatedEntries: { [key: number]: boolean } = {};
    const output: string[] = [];

    for (const line of lines) {
        if (line.startsWith('#')) {
            output.push(line);
            continue;
        }

        const [i18nKey, _] = line.split(/[=:]/).map((word) => word.trim());
        const index = entries.findIndex((entry) => entry.key === i18nKey);
        let newLine = line;

        if (index !== -1) {
            const { key, value } = entries[index];
            newLine = `${key}=${value}`;
            updatedEntries[index] = true;
        }
        output.push(newLine);
    }

    return { lines, updatedEntries, output };
}
