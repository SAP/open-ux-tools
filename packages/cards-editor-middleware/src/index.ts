import type { RequestHandler, Request, Response } from 'express';
import type { MiddlewareParameters } from '@ui5/server';
import { json, Router as createRouter } from 'express';
import path, { join } from 'path';
import { promises } from 'fs';
import { getWebappPath, FileName } from '@sap-ux/project-access';
import { render } from 'ejs';
import * as utils from './utilities';
import os from 'os';

export const enum ApiRoutes {
    previewGeneratorSandbox = '/test/flpGeneratorSandbox.html',
    cardsStore = '/cards/store',
    i18nStore = '/editor/i18n'
}

/**
 * Check if a file exists.
 *
 * @param path - The path to the file
 * @returns - true if the file exists, false otherwise
 */
async function pathExists(path: string) {
    try {
        await promises.access(path);
        return true;
    } catch (error) {
        if (error.code === 'ENOENT') {
            return false; // File does not exist
        } else {
            throw error; // Other errors
        }
    }
}

module.exports = async ({ resources }: MiddlewareParameters<any>): Promise<RequestHandler> => {
    const router = createRouter();
    router.use(json());

    const manifest = await resources.rootProject.byPath(`/${FileName.Manifest}`);
    if (!manifest) {
        throw new Error('manifest.json not found.');
    }

    /**
     * Route to serve card generator flp sandbox
     */
    router.get(ApiRoutes.previewGeneratorSandbox, async (_req, res: Response) => {
        const app = JSON.parse(await manifest.getString())['sap.app'];
        res.status(200).send(
            render(await promises.readFile(join(__dirname, '../templates/editor.html'), 'utf-8'), {
                templateModel: {
                    appTitle: app.title || 'Card Editor Preview',
                    component: app.id
                }
            })
        );
    });

    /**
     * Route to store card manifest files, the files are stored in the webapp folder of the project
     * and the application manifest.json file is updated with the new card manifests information within the sap.cards.ap.embeds
     */
    router.post(ApiRoutes.cardsStore, async (req: Request, res: Response) => {
        try {
            const { floorplan, localPath, fileName = FileName.Manifest, manifests } = req.body;
            const webappPath = await getWebappPath(path.resolve());
            const fullPath = path.resolve(webappPath, localPath);

            if (!fullPath.startsWith(webappPath)) {
                throw new Error('Invalid path');
            }

            const file = utils.prepareFileName(fullPath + '/' + fileName);
            if (!(await pathExists(fullPath))) {
                try {
                    await promises.mkdir(fullPath, { recursive: true });
                } catch (err) {
                    res.status(403).send(`Files could not be created/updated.`);
                }
            }

            const multipleCards = utils.prepareCardTypesForSaving(manifests);

            await promises.writeFile(join(fullPath, file), multipleCards.integration);
            await promises.writeFile(join(fullPath, 'adaptive-' + file), multipleCards.adaptive);

            const oManifest = JSON.parse(await manifest.getString());

            if (!oManifest['sap.cards.ap']) {
                oManifest['sap.cards.ap'] = {};
            }
            const integrationCard = manifests.find((card: any) => card.type === 'integration');
            const entitySet = integrationCard.entitySet;
            oManifest['sap.cards.ap'].embeds ??= {};
            oManifest['sap.cards.ap'].embeds[floorplan] ??= {
                default: entitySet,
                manifests: {}
            };
            oManifest['sap.cards.ap'].embeds[floorplan]['default'] ??= entitySet;
            oManifest['sap.cards.ap'].embeds[floorplan]['manifests'] ??= {};
            oManifest['sap.cards.ap'].embeds[floorplan]['manifests'][entitySet] ??= [
                {
                    'localUri': localPath
                }
            ];

            const manifestPath = join(webappPath, FileName.Manifest);
            await promises.writeFile(manifestPath, JSON.stringify(oManifest, null, 2));
            res.status(201).send(`Files were updated/created`);
        } catch (err) {
            res.status(500).send(`Files could not be created/updated.`);
        }
    });

    /**
     * Route to store i18n properties
     * All the new properties are added at the end of the i18n file
     */
    router.post(ApiRoutes.i18nStore, async (req: Request, res: Response) => {
        try {
            const oManifest = JSON.parse(await manifest.getString());
            const webappPath = await getWebappPath(path.resolve());
            let filePath: string;

            if (oManifest['sap.app'].i18n) {
                filePath = join(webappPath, oManifest['sap.app'].i18n);
            } else {
                filePath = join(webappPath, 'i18n', 'i18n.properties');
            }

            const entries = req.body || [];
            const { lines, updatedEntries, output } = await utils.traverseI18nProperties(filePath, entries);

            // Add a new line if file is not empty and last line is not empty and there are new entries
            if (lines?.length > 0 && lines[lines?.length - 1].trim() && entries?.length) {
                output.push('');
            }

            for (const index in entries) {
                const ikey = index as any;
                if (!updatedEntries[ikey]) {
                    const { comment, key, value } = entries[ikey];
                    if (comment) {
                        output.push(`#${comment}`); // Add comment only for a new entry
                    }
                    output.push(`${key}=${value}${os.EOL}`);
                }
            }

            await promises.writeFile(filePath, output.join(os.EOL), { encoding: 'utf8' });
            res.status(201).send(`i18n file updated.`);
        } catch (err) {
            res.status(500).send(`File could not be updated.`);
        }
    });

    return router;
};
