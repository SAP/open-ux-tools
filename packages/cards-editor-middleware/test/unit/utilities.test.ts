import type { I18nEntry } from '../../src/utilities';
import { prepareFileName, prepareCardTypesForSaving, traverseI18nProperties } from '../../src/utilities';
import { promises } from 'fs';
import packageJson from '../../package.json';

jest.mock('fs', () => ({
    promises: {
        ...jest.requireActual('fs').promises,
        readFile: jest.fn(),
        readdir: jest.fn(),
        stat: jest.fn()
    }
}));

describe('Common utilities', () => {
    const mockFsPromisesReadFile = promises.readFile as jest.Mock;
    beforeEach(() => {
        jest.resetAllMocks();
    });

    test('prepareFileName', () => {
        expect(prepareFileName('path/to/file.json')).toBe('file.json');
        expect(prepareFileName('path/to/file')).toBe('file.json');
    });

    test('prepareCardTypesForSaving', () => {
        const aMultipleCards = [
            {
                type: 'integration',
                manifest: {
                    '_version': '1.15.0',
                    'sap.card': {
                        'type': 'Object',
                        'header': {
                            'type': 'Numeric',
                            'title': 'Card title'
                        }
                    },
                    'sap.insights': {
                        'versions': {
                            'ui5': '1.121.0-202403281300'
                        },
                        'templateName': 'ObjectPage',
                        'parentAppId': 'sales.order.wd20',
                        'cardType': 'DT'
                    }
                }
            },
            {
                type: 'adaptive',
                manifest: {
                    'type': 'AdaptiveCard',
                    'body': [
                        {
                            'type': 'TextBlock',
                            'wrap': true,
                            'weight': 'Bolder',
                            'text': 'Card Title'
                        }
                    ]
                }
            }
        ];

        const expectedIntegrationCard = {
            '_version': '1.15.0',
            'sap.card': {
                'type': 'Object',
                'header': {
                    'type': 'Numeric',
                    'title': 'Card title'
                }
            },
            'sap.insights': {
                'versions': {
                    'ui5': '1.121.0-202403281300',
                    'dtMiddleware': packageJson.version
                },
                'templateName': 'ObjectPage',
                'parentAppId': 'sales.order.wd20',
                'cardType': 'DT'
            }
        };
        const preparedCards = prepareCardTypesForSaving(aMultipleCards);
        const integrationCard = JSON.parse(preparedCards.integration);

        expect(integrationCard).toMatchObject(expectedIntegrationCard);
    });

    test('traverseI18nProperties', async () => {
        const i18nContent = 'appTitle=Sales Order';
        mockFsPromisesReadFile.mockResolvedValueOnce(i18nContent);
        const entries: I18nEntry[] = [
            {
                'comment': 'XFLD: GroupPropertyLabel for new Entry - Created by Card Generator',
                'key': 'CardGeneratorGroupPropertyLabel_Groups_0_Items_0',
                'value': 'new Entry'
            }
        ];
        const { updatedEntries, output } = await traverseI18nProperties('path/to/i18n', entries);
        expect(updatedEntries).toEqual({});
        expect(output).toEqual([i18nContent]);
    });

    test('traverseI18nProperties, When new entry matches i18n file content', async () => {
        const i18nContent = 'appTitle=Sales Order';
        mockFsPromisesReadFile.mockResolvedValueOnce(i18nContent);
        const entries: I18nEntry[] = [{ 'key': 'appTitle', 'value': 'Sales Order' }];
        const { updatedEntries, output } = await traverseI18nProperties('path/to/i18n', entries);
        expect(updatedEntries).toEqual({ 0: true });
        expect(output).toEqual([i18nContent]);
    });
});
