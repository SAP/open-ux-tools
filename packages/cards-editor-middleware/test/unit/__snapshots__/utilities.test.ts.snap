// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common utils prepareCardTypesForSaving 1`] = `
Object {
  "adaptive": "{
  \\"type\\": \\"AdaptiveCard\\",
  \\"body\\": [
    {
      \\"type\\": \\"TextBlock\\",
      \\"wrap\\": true,
      \\"weight\\": \\"Bolder\\",
      \\"text\\": \\"Card Title\\"
    }
  ]
}",
  "integration": "{
  \\"_version\\": \\"1.15.0\\",
  \\"sap.card\\": {
    \\"type\\": \\"Object\\",
    \\"header\\": {
      \\"type\\": \\"Numeric\\",
      \\"title\\": \\"Card title\\"
    }
  },
  \\"sap.insights\\": {
    \\"versions\\": {
      \\"ui5\\": \\"1.120.1-202403281300\\",
      \\"dtMiddleware\\": \\"0.1.0\\"
    },
    \\"templateName\\": \\"ObjectPage\\",
    \\"parentAppId\\": \\"sales.order.wd20\\",
    \\"cardType\\": \\"DT\\"
  }
}",
}
`;
